File: Reqnroll/Chapter13/13.1-TheLocalScenarioContextPattern/After/WIMP.LocalScenarioContextSample.Specs/StepDefinitions/OrderingStepDefinitions.cs
using Reqnroll;

using WIMP.LocalScenarioContextSample.App.Infrastructure;
using WIMP.LocalScenarioContextSample.App.Services;

namespace WIMP.LocalScenarioContextSample.Specs.StepDefinitions;

[Binding]
public class OrderingStepDefinitions
{
    private int placedOrderNo;

    [Given("the customer {word} has placed the order #{int}")]
    public void GivenTheCustomerHasPlacedTheOrder(string customerName, int orderNo)
    {
        OrderService.PlaceOrder(customerName, orderNo, "Margherita");
        placedOrderNo = orderNo;  // Store in scenario context
    }

    [When("the customer {word} cancels the placed order")]
    public void WhenTheCustomerCancelsThePlacedOrder(string customerName)
    {
        OrderService.CancelOrder(customerName, placedOrderNo);  // Use stored order number
    }

    [Then("the customer {word} should receive a notification about the cancellation")]
    public void ThenTheCustomerShouldReceiveANotification(string customerName)
    {
        Assert.IsTrue(NotificationService.WasNotificationSent(customerName));
    }

    #region Reset database for every scenario execution

    /// <summary>
    /// This hook resets the in-memory database before each scenario execution,
    /// ensuring that each test starts with a clean state.
    /// The pattern TODO1 contains a better approach for in-memory database management via dependencies.
    /// The pattern TODO2 contains further options for dealing with shared resources, such as using a real database.
    /// </summary>
    [BeforeScenario]
    public void ResetDatabase()
    {
        DataContext.Instance.Reset();
    }

    #endregion
}

File: CucumberJVM/Chapter13/13.1-TheLocalScenarioContextPattern/Before/src/test/java/com/wimp/localscenariocontext/specs/stepdefinitions/OrderingStepDefinitions.java
package com.wimp.localscenariocontext.specs.stepdefinitions;

import com.wimp.localscenariocontext.app.infrastructure.DataContext;
import com.wimp.localscenariocontext.app.services.NotificationService;
import com.wimp.localscenariocontext.app.services.OrderService;
import io.cucumber.java.Before;
import io.cucumber.java.en.Given;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;

import static org.junit.jupiter.api.Assertions.assertTrue;

public class OrderingStepDefinitions {
    @Given("the customer {word} has placed the order #{int}")
    public void theCustomerHasPlacedTheOrder(String customerName, int orderNo) {
        OrderService.placeOrder(customerName, orderNo, "Margherita");
    }

    @When("the customer {word} cancels the order #{int}")
    public void theCustomerCancelsTheOrder(String customerName, int orderNo) {
        OrderService.cancelOrder(customerName, orderNo);
    }

    @Then("the customer {word} should receive a notification about the cancellation")
    public void theCustomerShouldReceiveANotification(String customerName) {
        assertTrue(NotificationService.wasNotificationSent(customerName));
    }

    /**
     * This hook resets the in-memory database before each scenario execution,
     * ensuring that each test starts with a clean state.
     * The pattern TODO1 contains a better approach for in-memory database management via dependencies.
     * The pattern TODO2 contains further options for dealing with shared resources, such as using a real database.
     */
    @Before
    public void resetDatabase() {
        DataContext.INSTANCE.reset();
    }
}
